File: macos_arm64/Sources/RimeoAgentMac/HTTPServer/JWTValidator.swift
import Foundation
import CryptoKit

// ES256 (ECDSA P-256 + SHA-256) JWT validator for protected endpoints
// exposed to the public internet through the named tunnel.
//
// Server-side signer is rimeo_webapp/jwt_signer.py (П3).
// Claims used here: aud (== this agent's tunnel hostname), exp.
//
// Public key is baked into the binary at release time. While the placeholder
// below is empty the validator no-ops and the agent stays usable for dev /
// quick-tunnel builds (graceful degrade — server also degrades when keys are
// not provisioned). П7 (agent_min_build gating) ensures clients only target
// agents whose build matches their authentication mode.
enum JWTValidator {
    // Contents of jwt_public.pem produced by tools/gen_jwt_keys.py
    // (SubjectPublicKeyInfo / PKIX PEM). The matching private key lives only in
    // the rimeo.app server env (JWT_PRIVATE_KEY_PEM / _PATH). Enforcement is
    // additionally gated to named-tunnel mode in APIRouter.jwtGate, so this key
    // stays inert while the agent is still on a quick tunnel.
    static let publicKeyPEM: String = """
    -----BEGIN PUBLIC KEY-----
    MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEUiawXlrjeQhLWfXjgs9FiuHjqOMt
    h0XfbBreh55YjgoyMsGys7emSXv3pBlUOeYWXGM3y+ZCiSPYuZNyHLWu2w==
    -----END PUBLIC KEY-----
    """

    enum Failure: String {
        case missing
        case malformed
        case wrongAlgorithm
        case expired
        case wrongAudience
        case invalidSignature
        case notConfigured

        var status: Int { self == .missing ? 401 : 401 }
    }

    private static let clockSkewSeconds: TimeInterval = 60
    private static let logLock = NSLock()
    private static var loggedNotConfigured = false

    static func isConfigured() -> Bool {
        !publicKeyPEM.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    static func validate(token rawToken: String?, expectedAudience: String) -> Failure? {
        guard isConfigured() else {
            logLock.lock()
            let shouldLog = !loggedNotConfigured
            loggedNotConfigured = true
            logLock.unlock()
            if shouldLog {
                logger.warning("JWT validation disabled: public key not configured (dev build / pre-rollout)")
            }
            return nil
        }
        guard expectedAudience.isEmpty == false else {
            // Named tunnel hostname unknown — refuse rather than trust a
            // token that may have been signed for a different agent.
            return .wrongAudience
        }
        guard let token = rawToken, !token.isEmpty else { return .missing }

        let parts = token.split(separator: ".", omittingEmptySubsequences: false)
        guard parts.count == 3 else { return .malformed }
        let headerB64  = String(parts[0])
        let payloadB64 = String(parts[1])
        let sigB64     = String(parts[2])

        guard let headerData  = base64URLDecode(headerB64),
              let payloadData = base64URLDecode(payloadB64),
              let sigData     = base64URLDecode(sigB64) else { return .malformed }

        guard let header = (try? JSONSerialization.jsonObject(with: headerData)) as? [String: Any],
              let alg = header["alg"] as? String else { return .malformed }
        if alg != "ES256" { return .wrongAlgorithm }

        guard let payload = (try? JSONSerialization.jsonObject(with: payloadData)) as? [String: Any] else {
            return .malformed
        }

        let expValue: Double?
        if let n = payload["exp"] as? Double { expValue = n }
        else if let n = payload["exp"] as? Int { expValue = Double(n) }
        else if let s = payload["exp"] as? String, let n = Double(s) { expValue = n }
        else { expValue = nil }
        guard let exp = expValue else { return .malformed }
        if Date().timeIntervalSince1970 > exp + clockSkewSeconds { return .expired }

        guard let aud = payload["aud"] as? String, !aud.isEmpty else { return .wrongAudience }
        if !audienceMatches(aud, expected: expectedAudience) { return .wrongAudience }

        let trimmedPEM = publicKeyPEM.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let pubKey = try? P256.Signing.PublicKey(pemRepresentation: trimmedPEM) else {
            return .notConfigured
        }
        guard let signature = try? P256.Signing.ECDSASignature(rawRepresentation: sigData) else {
            return .invalidSignature
        }
        let signingInput = Data((headerB64 + "." + payloadB64).utf8)
        let digest = SHA256.hash(data: signingInput)
        if !pubKey.isValidSignature(signature, for: digest) {
            return .invalidSignature
        }
        return nil
    }

    static func extractToken(from req: HTTPRequest) -> String? {
        if let q = req.queryParams["token"], !q.isEmpty { return q }
        if let auth = req.headers["authorization"], !auth.isEmpty {
            let lower = auth.lowercased()
            if lower.hasPrefix("bearer ") {
                return String(auth.dropFirst(7)).trimmingCharacters(in: .whitespaces)
            }
            return auth
        }
        return nil
    }

    private static func audienceMatches(_ aud: String, expected: String) -> Bool {
        if aud == expected { return true }
        // Tolerate scheme/host shapes: token may carry the bare hostname while
        // expected may be derived from a URL ("https://host"), or vice versa.
        let a = stripScheme(aud)
        let e = stripScheme(expected)
        return !a.isEmpty && a == e
    }

    private static func stripScheme(_ s: String) -> String {
        var v = s
        for prefix in ["https://", "http://"] {
            if v.hasPrefix(prefix) { v = String(v.dropFirst(prefix.count)) }
        }
        if let slash = v.firstIndex(of: "/") { v = String(v[..<slash]) }
        return v
    }

    private static func base64URLDecode(_ s: String) -> Data? {
        var str = s.replacingOccurrences(of: "-", with: "+")
                   .replacingOccurrences(of: "_", with: "/")
        let rem = str.count % 4
        if rem > 0 { str += String(repeating: "=", count: 4 - rem) }
        return Data(base64Encoded: str)
    }
}

File: windows_csharp/RimeoAgent/HttpServer/JwtValidator.cs
using System.Globalization;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace RimeoAgent.HttpServer;

/// ES256 (ECDSA P-256 + SHA-256) JWT validator for endpoints exposed to the
/// public internet through the named tunnel. Faithful port of the macOS
/// JWTValidator.swift; the server signer is rimeo_webapp/jwt_signer.py.
/// Claims used: aud (== this agent's tunnel hostname), exp.
///
/// Enforcement is gated to named-tunnel mode in ApiRouter (NamedHostname
/// non-empty), so this stays inert on quick/local tunnels where the server
/// does not sign tokens.
public static class JwtValidator
{
    public enum Failure
    {
        Missing,
        Malformed,
        WrongAlgorithm,
        Expired,
        WrongAudience,
        InvalidSignature,
        NotConfigured,
    }

    // SubjectPublicKeyInfo / PKIX PEM. The matching private key lives only in the
    // rimeo.app server env (JWT_PRIVATE_KEY_PEM). Same key baked into the macOS
    // build (JWTValidator.swift).
    private const string PublicKeyPem =
        "-----BEGIN PUBLIC KEY-----\n" +
        "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEUiawXlrjeQhLWfXjgs9FiuHjqOMt\n" +
        "h0XfbBreh55YjgoyMsGys7emSXv3pBlUOeYWXGM3y+ZCiSPYuZNyHLWu2w==\n" +
        "-----END PUBLIC KEY-----";

    private const int ClockSkewSeconds = 60;

    public static bool IsConfigured => !string.IsNullOrWhiteSpace(PublicKeyPem);

    public static string FailureReason(Failure f) => f switch
    {
        Failure.Missing          => "missing",
        Failure.Malformed        => "malformed",
        Failure.WrongAlgorithm   => "wrongAlgorithm",
        Failure.Expired          => "expired",
        Failure.WrongAudience    => "wrongAudience",
        Failure.InvalidSignature => "invalidSignature",
        Failure.NotConfigured    => "notConfigured",
        _                        => "unknown",
    };

    /// Returns null when the token is valid, or a Failure describing why not.
    /// Never throws — parse/crypto errors map to a Failure.
    public static Failure? Validate(string? rawToken, string expectedAudience)
    {
        if (!IsConfigured) return null;                              // dev / pre-rollout: no-op
        if (string.IsNullOrEmpty(expectedAudience)) return Failure.WrongAudience;
        if (string.IsNullOrEmpty(rawToken)) return Failure.Missing;

        var parts = rawToken.Split('.');
        if (parts.Length != 3) return Failure.Malformed;

        byte[] headerData, payloadData, sigData;
        try
        {
            headerData  = Base64UrlDecode(parts[0]);
            payloadData = Base64UrlDecode(parts[1]);
            sigData     = Base64UrlDecode(parts[2]);
        }
        catch { return Failure.Malformed; }

        try
        {
            using var headerDoc = JsonDocument.Parse(headerData);
            if (!headerDoc.RootElement.TryGetProperty("alg", out var algEl)
                || algEl.GetString() is not { } alg)
                return Failure.Malformed;
            if (alg != "ES256") return Failure.WrongAlgorithm;

            using var payloadDoc = JsonDocument.Parse(payloadData);
            var root = payloadDoc.RootElement;

            if (!TryGetExp(root, out double exp)) return Failure.Malformed;
            var now = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            if (now > exp + ClockSkewSeconds) return Failure.Expired;

            if (!root.TryGetProperty("aud", out var audEl)
                || audEl.GetString() is not { Length: > 0 } aud)
                return Failure.WrongAudience;
            if (!AudienceMatches(aud, expectedAudience)) return Failure.WrongAudience;

            using var ecdsa = ECDsa.Create();
            try { ecdsa.ImportFromPem(PublicKeyPem); }
            catch { return Failure.NotConfigured; }

            // JWT ES256 signature is raw r||s (fixed-width) = IEEE P1363.
            var signingInput = Encoding.ASCII.GetBytes(parts[0] + "." + parts[1]);
            bool ok = ecdsa.VerifyData(
                signingInput, sigData,
                HashAlgorithmName.SHA256,
                DSASignatureFormat.IeeeP1363FixedFieldConcatenation);
            return ok ? null : Failure.InvalidSignature;
        }
        catch { return Failure.Malformed; }
    }

    public static string? ExtractToken(AgentRequest req)
    {
        if (req.QueryParams.TryGetValue("token", out var q) && !string.IsNullOrEmpty(q))
            return q;
        if (req.Headers.TryGetValue("authorization", out var auth) && !string.IsNullOrEmpty(auth))
        {
            if (auth.StartsWith("bearer ", StringComparison.OrdinalIgnoreCase))
                return auth.Substring(7).Trim();
            return auth;
        }
        return null;
    }

    private static bool TryGetExp(JsonElement payload, out double exp)
    {
        exp = 0;
        if (!payload.TryGetProperty("exp", out var e)) return false;
        switch (e.ValueKind)
        {
            case JsonValueKind.Number:
                return e.TryGetDouble(out exp);
            case JsonValueKind.String:
                return double.TryParse(e.GetString(), NumberStyles.Any,
                    CultureInfo.InvariantCulture, out exp);
            default:
                return false;
        }
    }

    private static bool AudienceMatches(string aud, string expected)
    {
        if (aud == expected) return true;
        // Tolerate scheme/host shapes: token may carry the bare hostname while
        // expected is derived from a URL ("https://host"), or vice versa.
        var a = StripScheme(aud);
        var e = StripScheme(expected);
        return a.Length > 0 && a == e;
    }

    private static string StripScheme(string s)
    {
        var v = s;
        foreach (var prefix in new[] { "https://", "http://" })
            if (v.StartsWith(prefix, StringComparison.OrdinalIgnoreCase))
                v = v.Substring(prefix.Length);
        var slash = v.IndexOf('/');
        if (slash >= 0) v = v.Substring(0, slash);
        return v;
    }

    private static byte[] Base64UrlDecode(string s)
    {
        var str = s.Replace('-', '+').Replace('_', '/');
        switch (str.Length % 4)
        {
            case 2: str += "=="; break;
            case 3: str += "=";  break;
        }
        return Convert.FromBase64String(str);
    }
}
